src/lib.rs

```diff
@@ -1,129 +1,71 @@
-    // 3.   acc: `==`    ; next: `100`      ; tail: `%, >=1, >=1`         ; will match this rule again
-    //      acc: `==100` ;                    tail: `%, >=1, >=1`         ; is the next parse
-    // OR   acc: `==100` ; head: `%`; `,`;    tail: `>=1, >=1`            ; i.e. this is the match for the next parse
-    //                                ^^^ --------------------------------> this will match previous rule because of this comma
-    (@parse ($($acc:tt)*) $next:tt $($tail:tt)*) => {
-        $crate::constraints!(@parse ($($acc)* $next) $($tail)*)
-    };
-    // At the end there will a set of tokens left after the last `,`
-    // Process that as a `Constraint`
-    (@parse ($($acc:tt)*)) => {
-        [$crate::constraints!(@process ($($acc)*))]
-    };
-    (@process (== $token1:tt / $token2:tt ; $count:expr)) => {
-        // Percentage constraint
-        std::iter::repeat(ratatui::prelude::Constraint::Ratio($token1 as u32, $token2 as u32)).take($count as usize)
-    };
-    // Process different types of constraints into a `Constraint`
-    (@process (== $token1:tt / $token2:tt)) => {
-        // Ratio constraint
-        ratatui::prelude::Constraint::Ratio($token1 as u32, $token2 as u32)
-    };
-    (@process (== $token:tt % ; $count:expr)) => {
-        // Percentage constraint
-        std::iter::repeat(ratatui::prelude::Constraint::Percentage($token)).take($count as usize)
-    };
-    (@process (== $token:tt %)) => {
-        // Percentage constraint
-        ratatui::prelude::Constraint::Percentage($token)
-    };
-    (@process (>= $token:expr; $count:expr)) => {
-        // Percentage constraint
-        std::iter::repeat(ratatui::prelude::Constraint::Min($token)).take($count as usize)
+    // Note: this implementation forgoes speed for the sake of simplicity. Adding variations of the
+    // comma and semicolon rules for each constraint type would be faster, but would result in a lot
+    // of duplicated code.
+
+    // Comma finishes a constraint element, so parse it and continue.
+    ([ , $($rest:tt)* ] -> ($($partial:tt)*) [ $($parsed:tt)* ]) => {
+        $crate::constraints!([$($rest)*] -> () [$($parsed)* $crate::constraint!($($partial)*) ,])
     };
-    (@process (>= $token:expr)) => {
-        // Minimum size constraint
-        ratatui::prelude::Constraint::Min($token)
+
+    // Semicolon indicates that there's repetition. The trailing comma is because the 'entrypoint'
+    // rule adds a trailing comma.
+    ([ ; $count:expr , ] -> ($($partial:tt)*) []) => {
+        [$crate::constraint!($($partial)*); $count]
     };
-    (@process (<= $token:expr; $count:expr)) => {
-        // Percentage constraint
-        std::iter::repeat(ratatui::prelude::Constraint::Max($token)).take($count as usize)
+
+    // User wrote something like `constraints!([== 3, == 4; 10])`.
+    ([ ; $count:expr , ] -> ($($partial:tt)*) [$($_:expr)+]) => {
+        compile_error!("constraint repetition requires exactly one constraint")
```

Comment:
I'm adding a test suite now for fails but I'm not able to trigger this `compile_error!`.

Reply:
I've removed this rule since it wasn't getting triggered.